### PR TITLE
feat(forge): add cheatcodes vm.recordLogs() and vm.getRecordedLogs()

### PR DESCRIPTION
**Pull request by bentobox19**
## Overview

The methods `vm.recordLogs()` and `vm.getRecordedLogs()` provide the tester the ability to leverage information from the events emitted during the execution of a transaction.

Ref: https://github.com/foundry-rs/foundry/issues/543

## Motivation

While porting solutions from the CTF Ethernaut to Foundry, got bumped by the fact that when you create a challenge, you are calling the `Ethernaut.sol` contract with the address of the factory of the challenge you want to solve. `Ethernaut.sol` will invoke the factory, deploy a new challenge for you, and return the address of this new contract within an event `LevelInstanceCreatedLog(address indexed player, address instance)`.

In practical terms, one [can just modify the Ethernaut contract](https://github.com/OpenZeppelin/ethernaut/blob/60ba307bf3bc79c93b271d58c584b3905cf848b2/contracts/contracts/Ethernaut.sol#L48) to return you the address instead of emitting such event, _but_, this is a motivation to just build the feature into Foundry!

## Solution

* Implement `vm.captureLogs()`: Function mirroring `vm.record()`.
* Implement `vm.getRecordedLogs()`: Function mirroring `vm.accesses()`.

### CODE REVIEW

File: testdata/cheats/CaptureLogs.t.sol
Comment (onbjerg) [inline testdata/cheats/CaptureLogs.t.sol:83]: I wonder if this cheatcode should actually revert if there is no log, since I would assume you would only ever call it if you expect a log to be present

File: testdata/cheats/Cheats.sol
Comment (onbjerg) [inline testdata/cheats/Cheats.sol:66]: What's the primary reasoning for this cheatcode vs. something like `vm.getLogs()` and popping/indexing off of the returned array?

**Review by onbjerg — COMMENTED** [review]
Small comments, thanks for taking this on :smile:

File: testdata/cheats/CaptureLogs.t.sol
Comment (bentobox19) [reply testdata/cheats/CaptureLogs.t.sol:83]: Can be done.Yes.

File: testdata/cheats/Cheats.sol
Comment (bentobox19) [reply testdata/cheats/Cheats.sol:66]: 0. Hey @onbjerg I really appreciate your early review of this draft!
1. I'll do my best to be succint.
2. Disclaimer I'll do once: I have less than a week of experience in the marriage of Rust+Solidity 😅 !!! There are many features or patterns I ignore, and I'm most likely to propose stuff you already discussed and discarded. But well, there's only one way to start any jouney.

### Background

* Initially **I wanted something very simple™** and give the tester all the parsing work. After all, what I wanted as a feature is the ability to capture logs, leaving the UX problem as a future issue.

### Analysis

* What do we give to the tester if we invoke `vm.getLogs`?
   * Should it be a `bytes32[][]`. Maybe... But wait! `data` is not necessarily constained to 32 bytes!
   * So you have somehow an array of `(bytes32[], bytes)`. Can we do that? 🤔 
   * If we can't do above, can we just take the content of a `RawLog`, do a `(bytes32[], bytes)` tuple, `abi.encode` it into a `byte` field and then return a `byte[]` ? Seems cumbersome... Why don't we just go low-level and get the first log of the list in the simplest format we can afford?
* So in the end, I ended up thinking in UX!

### Proposal

* Having as cheatcodes `vm.captureLogs()` and `vm.nextCapturedLog()` within `forge`.
    * Then you just, in a FIFO fashion, **take** the first log from the list, leaving the next one in queue.
    * `vm.nextCapturedLog()` gives you back a `(bytes32[], bytes)`. The tester needs to do some quick decoding such as `abi.decode(data, (string))`.
    * Still working on how to work the Topic 0. For now its processing is to be at the tester discretion.
* Create later some wrapper at `forge-std` to `getLogs()`. This or these function(s) would be iterating over `nextCapturedLog`, taking care on indexing and decoding aspects.

### Wrapping up

My final goal is to be able to port the `Ethernaut` challenge into `forge`. I need the feature of capturing (and then reading those captured) logs. Let's take advantage we currently have some capacity on implementing this, and let's look towards the best solution 🙂

**Comment by bentobox19** [comment]
Hey @onbjerg I went a bit verbose here: https://github.com/foundry-rs/foundry/pull/2161#discussion_r912155983

Really open to discuss it until we converge into a solution. So far, it's been a very instructive experience where you have to reason in Rust, and low level solidity all in a single package.

File: evm/src/executor/abi.rs
Comment (mds1) [inline evm/src/executor/abi.rs:53]: Thanks for working on this PR! Definitely interested in it but haven't looked closely yet. Only comment for now is: should we keep the nomenclature consistent with `vm.record`, and therefore call this `vm.recordLogs`? 

I think using "captured" for the other cheatcode makes sense since these are not slot accesses and therefore `access` terminology doesn't quite fit

Comment (mds1) [inline evm/src/executor/abi.rs:53]: Thanks for working on this PR! Definitely interested in it but haven't looked closely yet. Only comment for now is: should we keep the nomenclature consistent with `vm.record`, and therefore call this `vm.recordLogs`? 

I think using "captured" for the other cheatcode makes sense since these are not slot accesses and therefore `access` terminology doesn't quite fit

File: testdata/cheats/Cheats.sol
Comment (mds1) [reply testdata/cheats/Cheats.sol:66]: Thanks for this info! I had the same thought as @onbjerg when looking at this. IMO I think the best UX would be something like this:
```solidity
interface Vm {
  // Records all transaction logs so they can be retrieved later.
  function recordLogs() external;

  // Returns an array of the recorded logs. `accessLog` is consistent with `accesses`, even
  // though semantics are a bit different. In general I don't feel strongly about all the cheatcode names 
  function accessLogs() external returns ((bytes32 memory topics, bytes memory data)[]);

  // Similar alternative.
  struct Log {
    bytes32[] topics;
    bytes data;
  }
  function logs() external returns (Log[] memory);
}
```

I don't personally see a big need for a cheatcode to access the next log, as opposed to just returning the full array and having the user loop through it or access a particular index. Providing the whole array instead is also nice because you if you know the index of an expected event it's cleaner to access it.

Comment (onbjerg) [reply testdata/cheats/Cheats.sol:66]: Thanks for the write-up @bentobox19! Don't worry about being new to Rust + Solidity, we'll help you if you need :smile: Looks good so far.

I agree with @mds1's 2nd proposal here: it would be entirely feasible (and imo preferable) to return `Log` structs. The nice thing about having it return *all* logs it captured between `recordLogs` and `logs` (or `accessLogs`, not sure what the best name is) is that helper "cheatcodes" like `nextLog` can be built on top in Forge Std or other libraries :smile: 

So I would definitely propose that you:

1. Remove the `nextCapturedLog` cheatcode
2. Replace it with a `logs` or `accessLogs` (still not sure about naming) that returns `Log[]`

A note on behavior of the `logs` cheatcode: I think it should return all logs between the call to `recordLogs` and the call to `logs`. After this the internal buffer should be cleared - in other words, the combination of these two cheatcodes define the range in which the logs were emitted, if that makes sense?

Comment (bentobox19) [reply testdata/cheats/Cheats.sol:66]: Thank you @mds1 and @onbjerg onbjerg for your prompt feedback: I am renaming to

```solidity
    // Record all the transaction logs
    function recordLogs() external;
    // Gets all the recorded logs
    function getRecordedLogs() external returns ((bytes32[] memory topics, bytes memory data)[]);
```

Let's pass on `getLogs()`, to avoid any potential confusion of the testers with the RPC method [eth_getLogs](https://docs.infura.io/infura/networks/ethereum/json-rpc-methods/eth_getlogs) 🤷‍♂️

**Comment by bentobox19** [comment]
In the end I went with

```solidity
    struct Log {bytes32[] topics; bytes data;}
    function recordLogs() external;
    function getRecordedLogs() external returns (Log[] memory);
```

The compilers weren't very friendly with

```solidity
    function getRecordedLogs() external returns ((bytes32[] memory topics, bytes memory data)[]);
```

So, the common flow is, for example

```solidity
        cheats.recordLogs();
        emitter.emitEvent(1, 2, 3, "some data");
        Cheats.Log[] memory entries = cheats.getRecordedLogs();

        assertEq(entries.length, 1);
        assertEq(entries[0].topics.length, 4);
        assertEq(entries[0].topics[0], keccak256("LogTopic123(uint256,uint256,uint256,bytes)"));
        assertEq(entries[0].topics[1], bytes32(uint256(1)));
        assertEq(entries[0].topics[2], bytes32(uint256(2)));
        assertEq(entries[0].topics[3], bytes32(uint256(3)));
        assertEq(abi.decode(entries[0].data, (string)), string("some data"));
```

On the last assertion, even if the data is not a `string`, but a `bytes` variable, `assertEq` wants to work with the former.

## The PR!

Alright. I'm setting this PR on:

* I'm satisfied with the working of both cheatcodes, I was driven by the tests at `RecordLogs.t.sol`.
* One of the tests uncovered a bug on [anonymous events](https://docs.soliditylang.org/en/v0.8.15/contracts.html#events). I'd like this to be filled as a separate issue.

Thank you.

**Comment by gakonst** [comment]
@bentobox19 this looks good - can you explain why we panic on anonymous events?

Defer to @mds1 on the interface being exposed, I like it

**Comment by bentobox19** [comment]
Hey @gakonst 

> can you explain why we panic on anonymous events?

This is my output when runing the test

```bash
[ bentobox: testdata ]$ ../target/debug/forge test -c cheats/RecordLogs.t.sol --match-test testRecordOnAnonymousEvent -vvvv
[⠑] Compiling...
No files changed, compilation skipped

Running 1 test for cheats/RecordLogs.t.sol:RecordLogsTest
[PASS] testRecordOnAnonymousEvent() (gas: 46536)
The application panicked (crashed).
Message:  index out of bounds: the len is 0 but the index is 0
Location: /Users/bentobox/Documents/dev/playground/archive/foundry/evm/src/trace/decoder.rs:299

This is a bug. Consider reporting it at https://github.com/foundry-rs/foundry

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

**Comment by bentobox19** [comment]
@gakonst 

I filled an issue with the panic bug here: https://github.com/foundry-rs/foundry/issues/2206

Anonymous events fail at verbosity 4+.

File: evm/src/executor/inspector/cheatcodes/mod.rs
Comment (onbjerg) [inline evm/src/executor/inspector/cheatcodes/mod.rs:78]: Might be better to have a boolean indicating whether we are recording or not so you can do something like:

```solidity
vm.recordLogs();
// some calls (lets say call a, b and c)
vm.getRecordedLogs(); // gets logs for calls a, b and c
// perform call d and call e
vm.getRecordedLogs(); // gets logs for calls d and e
```

Additionally I think a call to `vm.getRecordedLogs` should only return logs between the call to `getRecordedLogs` and the previous call to `recordLogs` or `getRecordedLogs` to allow for the pattern above. Currently this is the behavior:

```solidity
vm.recordLogs();
// some calls (lets say call a, b and c)
vm.getRecordedLogs(); // gets logs for calls a, b and c
// perform call d and call e
vm.getRecordedLogs(); // gets logs for  calls a, b, c, d and e
```

To discard logs for the first 3 calls you would have to call `recordLogs` again to reset the buffer

File: evm/src/executor/inspector/cheatcodes/env.rs
Comment (onbjerg) [inline evm/src/executor/inspector/cheatcodes/env.rs:116]: Imo this should clear the `state.recorded_logs` vec (see reasoning below)

File: evm/src/executor/inspector/cheatcodes/env.rs
Comment (onbjerg) [inline evm/src/executor/inspector/cheatcodes/env.rs:124]: `Vec<H256>` and `Vec<u8>` implement the `Tokenizable` trait, which also means it implements the `Tokenize` trait, so I think it should be possible to just call `entry.topics.into_tokens()` and `entry.data.into_tokens()`?

**Review by onbjerg — CHANGES_REQUESTED** [review]
Small nits. Regarding my comments on the API/behavior, I would give @mds1 final word on it since he is more in touch with actually using it (I don't write much Solidity these days)

**Comment by bentobox19** [comment]
Thank you for your feedback @onbjerg. We wait for @mds1 comments then?

**Comment by mds1** [comment]
@bentobox19 The proposed API looks great to me, thank you! 👌

**Comment by bentobox19** [comment]
Wait, @mds1, should I work on the API proposal of @onbjerg then?

**Comment by mds1** [comment]
Yea to clarify I think the below VM interfaces look good:

```solidity
struct Log {bytes32[] topics; bytes data;}
function recordLogs() external;
function getRecordedLogs() external returns (Log[] memory);
```

And I think the above suggestion of behavior from @onbjerg is good too:

```solidity
vm.recordLogs();
// some calls (lets say call a, b and c)
vm.getRecordedLogs(); // gets logs for calls a, b and c
// perform call d and call e
vm.getRecordedLogs(); // gets logs for calls d and e
```

**Comment by gakonst** [comment]
Great - let's address the @onbjerg code review comments, and given we like the API we can get it over the line!

File: evm/src/executor/inspector/cheatcodes/env.rs
Comment (bentobox19) [reply evm/src/executor/inspector/cheatcodes/env.rs:124]: Some hair pulling here, but managed to refashion this part of the code

```rs
        ethers::abi::encode(
            &recorded_logs
                .entries
                .iter()
                .map(|entry| {
                    Token::Tuple(vec![
                        entry.topics.clone().into_token(),
                        Token::Bytes(entry.data.clone()),
                    ])
                })
                .collect::<Vec<Token>>()
                .into_tokens(),
        )
        .into()
```

Sample output object:

> This maps to `(bytes32[], bytes)[]` or `Log[]` with `struct Log {bytes32[] topics; bytes data;}`

```rs
[ Tuple([
  Array([
    FixedBytes([97, 251, 125, ...]), 
    FixedBytes([... 0, 1])]), 
  Bytes([... 128, 64, 178, 137, 171, 233, 51, 52, 165, 227, ...])]), 
  Tuple([
    Array([
      FixedBytes([122, 249, 45, ...]), 
      FixedBytes([... 0, 2]), 
      FixedBytes([... 0, 3])]), 
    Bytes([... 171, 233, 51, 52, 165, 227, 102, 39, 81, 170, ...])]), 
...
```

* Using both `.into_token()` and `into_tokens()`.
* `into_tokens()` gives you as output a `[Array[ ....`
* `entry.data.clone().into_token()` will give you `Array([Uint(0), Uint(0),` instead of the desired `Bytes([0, 0, 0, 0, 0, `

File: evm/src/executor/inspector/cheatcodes/env.rs
Comment (bentobox19) [reply evm/src/executor/inspector/cheatcodes/env.rs:116]: Will copy `state.recorded_logs` and then call `start_record_logs()` to reset them.

**Comment by bentobox19** [comment]
@mds1 @onbjerg changes implemented.

File: evm/src/executor/inspector/cheatcodes/env.rs
Comment (onbjerg) [reply evm/src/executor/inspector/cheatcodes/env.rs:116]: You can also use `std::mem::take` instead

File: evm/src/executor/inspector/cheatcodes/env.rs
Comment (onbjerg) [inline evm/src/executor/inspector/cheatcodes/env.rs:119]:
```suggestion

```

**Review by onbjerg — CHANGES_REQUESTED** [review]
smallest nit (see https://doc.rust-lang.org/std/option/enum.Option.html#method.replace)

**Review by onbjerg — APPROVED** [review]
Rebased, looks good to merge when CI passes :smile: Great job!

**Comment by onbjerg** [comment]
Any final comments on API @gakonst / @mds1? Otherwise this can be merged :)

**Comment by mds1** [comment]
I think we're missing the new `Vm` methods in https://github.com/foundry-rs/foundry/blob/master/testdata/cheats/Cheats.sol, but otherwise LGTM!

Thank you @bentobox19! 🎉

**Review by gakonst — APPROVED** [review]
Awesome. Fantastic. Let's document it in the book?

**Comment by bentobox19** [comment]
@gakonst I can send a PR to the book repository today

**Comment by bentobox19** [comment]
@gakonst https://github.com/foundry-rs/book/pull/427

**Pull request merged**
